source/opt/value_number_table.cpp

```diff
@@ -16,29 +16,49 @@
 
 #include <algorithm>
 
+#include "cfg.h"
+
 namespace spvtools {
 namespace opt {
 
 uint32_t ValueNumberTable::GetValueNumber(spvtools::ir::Instruction* inst) {
-  // TODO: Need to implement the substitution of operands by their value number
-  // before hashing.
-  // TODO: Implement a normal form for opcodes that commute like integer
-  // addition.  This will let us know that a+b is the same value as b+a.
   assert(inst->result_id() != 0 &&
          "inst must have a result id to get a value number.");
 
   // Check if this instruction already has a value.
-  auto id_to_val = id_to_value_.find(inst->result_id());
-  if (id_to_val != id_to_value_.end()) {
-    return id_to_val->second;
+  auto result_id_to_val = id_to_value_.find(inst->result_id());
+  if (result_id_to_val != id_to_value_.end()) {
+    return result_id_to_val->second;
+  }
+  return 0;
+}
+
+uint32_t ValueNumberTable::AssignValueNumber(ir::Instruction* inst) {
+  // If it already has a value return that.
+  uint32_t value = GetValueNumber(inst);
+  if (value != 0) {
+    return value;
   }
 
   // If the instruction has other side effects, then it must
   // have its own value number.
+  // OpSampledImage and OpImage must remain in the same basic block in which
+  // they are used, because of this we will assign each one it own value number.
   if (!context()->IsCombinatorInstruction(inst)) {
-    uint32_t value_number = TakeNextValueNumber();
-    id_to_value_[inst->result_id()] = value_number;
-    return value_number;
+    value = TakeNextValueNumber();
```

Comment:
Why not just return an invalid value number? These special instructions (combinators, images, etc) seem to be treated differently than other things which don't have useful values. Below you refuse to assign a value to a phi if all the inputs are not the same, but you could assign a new value number. The two cases seem inconsistent.

Reply:
I always assign a value number to phis.  I want to assign a value number to every instruction.  The first reason is that it leads to fewer special cases: I don't have to look for the invalid value number.  Second, I don't know how people will use them, so a more general approach seems better.